packages/block-editor/src/components/block-controls/index.js

```diff
@@ -10,15 +20,32 @@ import { ifBlockEditSelected } from '../block-edit/context';
 
 const { Fill, Slot } = createSlotFill( 'BlockControls' );
 
-const BlockControlsFill = ( { controls, children } ) => (
-	<Fill>
-		<Toolbar controls={ controls } />
-		{ children }
-	</Fill>
-);
+function BlockControlsSlot( props ) {
+	const accessibleToolbarState = useContext( ToolbarContext );
+	return <Slot { ...props } fillProps={ accessibleToolbarState } />;
+}
+
+function BlockControlsFill( { controls, children } ) {
+	return (
+		<Fill>
+			{ ( fillProps ) => {
+				// Children passed to BlockControlsFill will not have access to any
+				// React Context whose Provider is part of the BlockControlsSlot tree.
+				// So we re-create the Provider in this subtree.
+				const value = ! isEmpty( fillProps ) ? fillProps : null;
+				return (
+					<ToolbarContext.Provider value={ value }>
```

Comment:
I remember having to do the same trick in the past for other fills. It's a bit unfortunate that we need this trick but as long it works it feels like a good trade-off.